src/walker.rs

```diff
@@ -0,0 +1,183 @@
+//!   Visiting a node includes processing it and visiting other related
+//!   entities, based on the node type.
+//!
+//! Process: As part of visiting a node, the node itself must be processed.
+//!   This is essentially calling back a provided function with the node as
+//!   argument.
+//!
+//! Modifications: At each node, the list of children or successor/predecessor
+//!   nodes, to be visited, is collected first, before any of them are visited.
+//!   So any modifications, insertions or deletions to the graph must take this
+//!   into account, to avoid visiting deleted nodes or missing visits to newly
+//!   inserted nodes.
+//!
+//! Termination: Cycles are possible b/w the blocks in a region.
+//!   It is the responsiblity of the [VisitBlock::blocks_visit_order] to ensure
+//!   termination. No cycle detection is done by the walker.
+
+use crate::{
+    basic_block::BasicBlock,
+    context::{Context, Ptr},
+    error::Result,
+    operation::Operation,
+    region::Region,
+};
+
+/// When the walker arrives at a node, it calls the Visit*
+/// callbacks, that return a list of [VisitAction], and does that.
+pub enum VisitAction<T> {
+    /// Process the node being visited.
+    Process,
+    /// Visit other nodes.
+    Visit(T),
+}
+
+/// When a walk arrives at a [BasicBlock], this trait decides when its
+/// successors / predecessors are visited, and when the block itself be processed.
+pub trait VisitBlock {
+    /// Given a block, determine when the successor/predecessor blocks are visited
+    /// and when the block itself is processed.
+    fn blocks_visit_order(
+        &mut self,
+        ctx: &Context,
+        cur_block: Ptr<BasicBlock>,
+    ) -> Vec<VisitAction<Ptr<BasicBlock>>>;
+
+    /// Given a block, get the list of [Operation]s in the block that must be visited.
+    fn operations_visit_order(
+        &mut self,
+        ctx: &Context,
+        cur_block: Ptr<BasicBlock>,
+    ) -> Vec<Ptr<Operation>>;
+}
+
+/// Callbacks when a [BasicBlock] is processed.
+pub trait ProcessBlock {
+    /// Call back before visiting the instructions in a [BasicBlock].
+    fn process_pre_insts(&mut self, _ctx: &mut Context, _block: Ptr<BasicBlock>) -> Result<()> {
+        Ok(())
+    }
+    /// Call back after visiting the instructions in a [BasicBlock].
+    fn process_post_insts(&mut self, _ctx: &mut Context, _block: Ptr<BasicBlock>) -> Result<()> {
+        Ok(())
+    }
+}
+
+/// When a walk arrives at a [Region], specify which of its blocks, and in what
+/// order, must be visited.
+pub trait VisitRegion {
+    /// Given a region, get a list of child blocks that must be visited.
+    /// Typically, this is just the entry block or the exit blocks since
+    /// the other blocks subsequently get visited via [VisitBlock::blocks_visit_order].
+    fn blocks_visit_order(
+        &mut self,
+        ctx: &Context,
+        cur_region: Ptr<Region>,
+    ) -> Vec<Ptr<BasicBlock>>;
+}
```

Comment:
The Visit interfaces makes it look like we potentially will have a large number of allocations to expect due to small vectors being created. I have had the same problem in the proposal when allowing `&mut Context` due to the borrow checker (rightfully). 

3 ways to reduce intermittent small allocations could either be
- return an iterator instead of vector (if possible?)
- use 'smallvec'
- allow the callee to add elements to the workset. For example by changing the interface to:

```
fn blocks_visit_order(&mut self, ctx: &Context, &mut workset: Workset<Ptr<BasicBlock>>, cur_block: Ptr<BasicBlock>)
```

In the later case it might be possible to keep the Workset around between calls such that adding more elements might not require new allocations.

Somehow this reminds me in common graph traversal algorithms where we keep a working set of nodes to visit. Depending on how we add the nodes to the workset we will get a different walking strategy. 

Just as a thought experiment, but I wonder what would happen if we really treat the complete IR as a graph and allow anyone to add any kind of IR object to the workset. e.g. have the workset be something like this:

```
type Workset = VecDequeue<Node>;

enum Node {
  Op(Ptr<Operation>)
  Block(Ptr<BasicBlock>)
  Region(Ptr<Region>)
}
```

That way one can schedule visits of node types arbitrarily.

Reply:
>Just as a thought experiment, but I wonder what would happen if we really treat the complete IR as a graph and allow anyone to add any kind of IR object to the workset. e.g. have the workset be something like this:

I like this idea very much, but that brings me to a more fundamental question. Do we need any of this walker/visitor infrastructure at all.

It's really just 3 loops that we're trying to provide all this interfaces for, and the effort to implement various callbacks / traits seem to be more than actually manually iterating over the IR. What's the downside to manual iteration without all this interfaces? On the other hand, every downside we've discussed in this PR and your proposal goes away with the user just manually writing the three loops, either directly as loops or mutually recursive functions, or even better: a single loop using the `worklist` method that you just suggested.